instrumentation/github.com/Shopify/sarama/otelsarama/producer.go

```diff
@@ -176,25 +183,31 @@ func WrapAsyncProducer(saramaConfig *sarama.Config, p sarama.AsyncProducer, opts
 					continue
 				}
 
-				span := startProducerSpan(cfg, saramaConfig.Version, msg)
+				span, ok := startProducerSpan(cfg, saramaConfig.Version, msg)
 
-				// Create message context, backend message metadata
-				mc := producerMessageContext{
-					metadataBackup: msg.Metadata,
-					span:           span,
-				}
+				var mc producerMessageContext
+				if ok {
+					// Create message context, backend message metadata
+					mc = producerMessageContext{
+						metadataBackup: msg.Metadata,
+						span:           span,
+					}
 
-				// Specific metadata with span id
-				msg.Metadata = span.SpanContext().SpanID()
+					// Specific metadata with span id
+					msg.Metadata = span.SpanContext().SpanID()
+				}
 
 				p.Input() <- msg
-				if saramaConfig.Producer.Return.Successes {
-					producerMessageContexts[msg.Metadata] = mc
-				} else {
-					// If returning successes isn't enabled, we just finish the
-					// span right away because there's no way to know when it will
-					// be done.
-					finishProducerSpan(span, msg.Partition, msg.Offset, nil)
+
+				if ok {
+					if saramaConfig.Producer.Return.Successes {
+						producerMessageContexts[msg.Metadata] = mc
+					} else {
+						// If returning successes isn't enabled, we just finish the
+						// span right away because there's no way to know when it will
+						// be done.
+						finishProducerSpan(span, msg.Partition, msg.Offset, nil)
```

Comment:
I think there is a bug in the conditional logic here.
should it not be something like:
```go
switch {
	case ok && saramaConfig.Producer.Return.Successes:
		producerMessageContexts[msg.Metadata] = mc
	case !saramaConfig.Producer.Return.Successes:
		span.End()
}
```